### PR TITLE
[js-api] Integrate with the ResizableArrayBuffer proposal

### LINKED ISSUE #1292: [js-api] Integration with the ResizableArrayBuffer and GrowableSharedArrayBuffer proposal
The [ResizableArrayBuffer and GrowableSharedArrayBuffer proposal](https://github.com/tc39/proposal-resizablearraybuffer) is a TC39 proposal that is currently stage 2.

There was [some discussion](https://github.com/tc39/proposal-resizablearraybuffer/issues/5) of how the proposal might integrate with the Wasm/JS API. It seems better to move the discussion here to get a concrete plan and to get more Wasm eyes on the issue.

As a starting point, I suggest the following changes to the WebAssembly.Memory API:

- `MemoryDescriptor` be amended with a `exposeAsResizableBuffer` boolean, defaulting to `false`.
- `Memory` will have a new `resizableBuffer` attribute.
    - If the instance is created with a descriptor whose `exposeAsResizableBuffer` is `true`, then the `buffer` attribute will throw a TypeError, `resizableBuffer` will return a `ResizableArrayBuffer` instance or a `GrowableSharedArrayBuffer` instance if it's a shared memory.
       - If the `Memory` was created with a maximum size, that size will be reflected on the `ResizableArrayBuffer` or `GrowableSharedArrayBuffer` instance.
       - Otherwise, some implementation-defined max size will be reflected.
    - If the instance is created with a descriptor whose `exposeAsResizableBuffer` is `false`, the the `resizableBuffer` attribute will throw a TypeError.
- `ResizableArrayBuffer`s gotten via `Memory`'s `buffer` attribute will:
    - Throw on `transfer()`, since they can't be detached.
    - Throw on `resize()` that are shrinks.

Thoughts welcome.

(Also, what is the process for bringing such a "normative PR" that's not a full proposal to the CG? Is it just an agenda item?)

### PR DESCRIPTION
**Pull request by syg**
Closes #1292 

This PR integrates the [resizable buffers proposal](https://github.com/tc39/proposal-resizablearraybuffer) with js-api.

Example use of the proposed API:

```javascript
// mem is a WebAssembly.Memory
let ab = mem.buffer;
assert(!ab.resizable);

// Make .buffer resizable, detaching the old one.
let rab = mem.toResizableBuffer();
assert(rab.resizable);
assert(mem.buffer === rab);
assert(%IsDetached(ab));

// We can go back if we choose.
let ab2 = mem.toFixedLengthBuffer();
assert(!ab2.resizable);
assert(mem.buffer === ab2);
assert(ab2 !== ab);
assert(%IsDetached(rab));

// Let's go back to resizable. Memory#grow no longer detaches .buffer when it's resizable.
rab = mem.toResizableBuffer();
let oldLen = rab.byteLength;
mem.grow(8);
assert(!%IsDetached(rab);
assert(rab.byteLength === oldLen + (8 * 65536))

// RAB#resize throws when trying to shrink or grow by non-page multiples for WebAssembly.Memory-vended RABs.
assertThrows(() => rab.resize(rab.byteLength - 65536), RangeError);
assertThrows(() => rab.resize(rab.byteLength + 10), RangeError);

### CODE REVIEW

**Review by littledan — DISMISSED** [review]
This proposed integration looks great to me. I like the API shape and the semantics of when and how rounding happens.

I'm happy to see proposals move through WebAssembly CG's process in the form of a PR, which seems more clear and lightweight than the whole-repository fork.

Have we heard feedback from SpiderMonkey or JSC about this PR? cc @lars-t-hansen @kmiller68

File: document/js-api/index.bs
Comment (syg) [reply document/js-api/index.bs:771]: No real reason. I was erring on giving implementations more latitude. I don't know of an actual use case to have this _not_ be the limit, so happy to change it.

File: document/js-api/index.bs
Comment (syg) [reply document/js-api/index.bs:779]: That's an interesting idea for hosts to write a method. It'll be a new editorial convention.

Do I understand the nit correctly that you see this as an implementation hook instead of a host hook, thus the suggestion?

**Review by syg — COMMENTED** [review]
Thanks for the review.

File: document/js-api/index.bs
Comment (syg) [reply document/js-api/index.bs]: Changed to 65536 × 65536.

**Comment by rossberg** [comment]
@syg, this PR is stale, do you still intend to land it?

**Comment by syg** [comment]
I do, yes, but it's not clear to me when I should be doing that. This PR is phase... 3, I think? How should integration PRs like this land? My recollection is that we're waiting for the upstream JS feature to get into the JS standard (i.e. stage 4), which hasn't happened yet.

**Comment by dtig** [comment]
We discussed this in the [6/22/21 meeting](https://github.com/WebAssembly/meetings/blob/main/main/2021/CG-06-22.md), though the meeting notes aren't very verbose. What I remember matches @syg's comment above that as this is an integration feature, we do need this to be standardized in TC39, before this is merged into the Wasm spec. The plan for this specifically is that once the JS feature lands, this comes back to the Wasm CG for a final poll, and can land with edits if needed.

**Comment by dtig** [comment]
I'd like some input on the usage patterns of this API in the tools. One of the [motivations](https://github.com/tc39/proposal-resizablearraybuffer#sync-up-capability-with-webassembly-memorygrow) of the ResizableArrayBuffer proposal is to expose memory.grow to JS without having to detach the existing ArrayBuffer. If a Wasm instance has a "default" buffer associated with it, to ensure that it doesn't detach on Grow it stands to reason that it should be a ResizableArrayBuffer. The change that this PR makes is that the buffer getter returns a fixed length ArrayBuffer (which makes sense for backwards compatibility). My question here is how would tools use this API? Is the expectation that applications would use toResizableBuffer in place of the buffer getter? 

On a side note, this proposal also needs a similar PR to the [threads](https://github.com/WebAssembly/threads) repository to integrate GSABs with shared Wasm memory.

**Comment by rossberg** [comment]
Has there been any progress on this?

**Comment by syg** [comment]
@rossberg My understanding is there's discussion on whether it's okay for the wasm spec to depend on JS proposals that aren't yet merged into ecma262. The status quo is no, so I've been conservative and letting this sit open until we get resizable buffers to stage 4 in TC39.

Currently the proposal is stage 3 in TC39. Chrome has [shipped in M111](https://chromestatus.com/feature/4668361878274048), and  I see Safari [enabled it in STP 160](https://developer.apple.com/safari/technology-preview/release-notes/). I can probably ask for stage 4 soon, but the practical upshot it's stable and shipped and there shouldn't be too much risk for the wasm spec to depend on it even during stage 3.

**Comment by rossberg** [comment]
@syg, makes sense, thanks for the update!

**Comment by syg** [comment]
@conrad-watt Could you give me an overview of what needs changing on the threads proposal to account for resizable buffers? (Or, if I'm really feeling lucky, are you up for making the changes?)

**Comment by conrad-watt** [comment]
I'm happy to take a look at this myself, but I'm currently working through a rebase of the threads repo (with thanks to @ioannad!) which I'd prefer to finish first. Is it ok if I get back to you in ~2 weeks?

Looking at the language in the current draft threads JS API, my knee-jerk feeling is that growable SharedArrayBuffer integration won't be too difficult an extension, so long as I can crib from this PR.

**Comment by syg** [comment]
@conrad-watt 2 weeks sgtm! I'm getting ready to ask for Stage 4 for resizable buffers so crossing my t's and dotting my i's.

**Comment by conrad-watt** [comment]
Ok, I have a fresh version of the threads specification now living in a [branch](https://github.com/WebAssembly/threads/tree/upstream-rebuild). @syg I'll spend a few days looking at growable SharedArrayBuffer and see what I can do (although I'll probably end up begging for help at some point).

Just as a very surface-level question, is the intention that growable buffers will be produced by `toResizableBuffer` on a shared memory, or will there be a separate `toGrowableBuffer` method for shared memories? I have a slight knee-jerk preference for the former.

**Comment by conrad-watt** [comment]
Just a note here that we should probably keep an eye on https://github.com/tc39/ecma262/pull/3116

File: document/js-api/index.bs
Comment (Ms2ger) [inline document/js-api/index.bs:748]: This is not great - I think we should have a global `HostResizeArrayBuffer` definition, which checks if its `buffer` argument is one of these special buffers - I guess we could even use `[[ArrayBufferDetachKey]]` for that

**Comment by bakkot** [comment]
The JS proposal has landed in the specification.

**Comment by ioannad** [comment]
@Ms2ger does the [last commit](https://github.com/WebAssembly/spec/pull/1300/commits/31964d4e1a052729741126eae013b2058a569d0e) address your review comment?

File: document/js-api/index.bs
Comment (Ms2ger) [inline document/js-api/index.bs:766]:
```suggestion
         1. Return <emu-const>handled</emu-const>.
    1. Otherwise, return <emu-const>unhandled</emu-const>.
```

File: document/js-api/index.bs
Comment (Ms2ger) [inline document/js-api/index.bs:764]:
```suggestion
                1. [=Grow the memory buffer=] associated with |memaddr| by |delta|.
```

**Comment by rossberg** [comment]
@syg, can you address @Ms2ger's remaining comments, so that we can merge this?

**Comment by syg** [comment]
I think @ioannad would better handle it, as she's been handling @Ms2ger's previous reviews. Sorry to pass the buck. :)

**Comment by castholm** [comment]
@Ms2ger I'm just an outside observer so apologies I'm stating the obvious, but I wanted to point out that this was merged into the `master` branch and not `main`, so as a result the linked issue was not closed and the updated spec was not deployed to https://webassembly.github.io/spec/.

**Comment by Ms2ger** [comment]
Argh. I opened #1742, but of course now I can't approve it, because I opened it. Thanks for flagging!

**Pull request merged**
